### DISCUSSION TITLE
How about passing codeblock content to widget with same name

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#6466] [Ideas]
In obsidian, some plugin can do

```md
```mermaid
xxx
/```
```

But in tid we have to

```jsx
<$mermaid text='
  graph
    A --> B
    B --> C
    C --> A
    D --> C
'></$mermaid>
```

If we can pass codeblock content to widget with the same name, we can do what obsidian can do.

**Comment by Jermolene** [comment]
Hi @linonetwo we can achieve the same effect in TiddlyWiki using custom parser and a typed code block. I don't think there is a mime type for Mermaid, so we'd have to make one up. All that is required is for the plugin to register a parser for the mime type.

```
$$$text/x-mermaid.js
  graph
    A --> B
    B --> C
    C --> A
    D --> C
$$$
```

I appreciate that that is not the same as the familiar fenced code block syntax. We could definitely explore more extensible handling for the optional "type" tag (at the moment, the highlight plugin just overwrites the core codeblock implementation with a custom one; we should consider allowing multiple plugins to register codeblock types.

**Reply by linonetwo** [reply]
Thanks, Jeremy.

Understood, I will try adding parser in my future plugins to explore this.

And I found this already works now (with tw-mermaid plugin)!

```md
$$$text/vnd.tiddlywiki.mermaid
  graph
    A --> B
    B --> C
    C --> A
    D --> C
$$$
```

And about the codeblock widget, this problem is pretty similar to https://github.com/Jermolene/TiddlyWiki5/discussions/6354#discussioncomment-1847350 . We can wait for a time to see if there are more feature requests about the widget base class.

Maybe we can have several hooks in the widget in the future, to make JS plugin developer like me being able to do more things...

**Reply by linonetwo** [reply]
> plugin to register a parser for the mime type.

For user's convenience, can I register a shorter type, like just `mermaid`?

Otherwise we may need https://github.com/Gk0Wk/TW5-CodeMirror-Enhanced to provide autocomplete of these types, let user being able to type `mermaid` and we autocomplete `text/vnd.tiddlywiki.mermaid` (Yes, https://efurlanm.github.io/mermaid-tw5.html is using this).

Hope there is an API to get the type list...

**Reply by Jermolene** [reply]
> And about the codeblock widget, this problem is pretty similar to https://github.com/Jermolene/TiddlyWiki5/discussions/6354#discussioncomment-1847350

Thank you, I'd forgotten about that ticket, but yes, it's the same topic.

> For user's convenience, can I register a shorter type, like just mermaid?

If the Mermaid plugin registers a file extension as well then you can use that instead of the mime type.

**Comment by linonetwo** [comment]
Update: Github can embed mermaid too! https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

```mermaid
  graph
    A --> B
    B --> C
    C --> A
    D --> C
```

```
```mermaid
  graph
    A --> B
    B --> C
    C --> A
    D --> C
/```
```
